### PR TITLE
ML/LlamaIndex: Add software tests and CI configuration

### PR DESCRIPTION
**Pull request by amotl**
## About
A followup on GH-691.
- Adding software tests and a CI configuration, to verify the example program does not break.
- Nightly runs will verify that the integration will always work, or signal when it breaks.
- The outcome will be reflected on the [Build Status](https://cratedb.com/docs/crate/clients-tools/en/latest/status.html) page.
- Dependabot will drive the project dependencies.

## References
- GH-691

### CODE REVIEW

File: topic/machine-learning/llama-index/init.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    ('2023-09-14T04:00:00', 17.3, 'Sensor B', 2),
+    ('2023-09-14T05:00:00', 20.1, 'Sensor B', 2),
+    ('2023-09-14T06:00:00', 22.5, 'Sensor A', 1),
+    ('2023-09-14T07:00:00', 18.3, 'Sensor A', 1),
+    ('2023-09-14T08:00:00', 16.8, 'Sensor A', 1),
+    ('2023-09-14T09:00:00', 14.6, 'Sensor B', 2),
+    ('2023-09-14T10:00:00', 13.2, 'Sensor B', 2),
+    ('2023-09-14T11:00:00', 11.7, 'Sensor B', 2);
+
+REFRESH TABLE time_series_data;
```
Comment (amotl) [inline topic/machine-learning/llama-index/init.sql:23]: The new `init.sql` file could also be used within the README to be provisioned by using an incantation to `crash`.

File: topic/machine-learning/llama-index/requirements.txt [hunk 2]
```diff
@@ -1,100 +1,7 @@
...
-verlib2==0.2.0
-wrapt==1.16.0
-yarl==1.16.0
+langchain-openai<0.3
+llama-index<0.12
+llama-index-embeddings-langchain<0.3
+llama-index-embeddings-openai<0.3
+llama-index-llms-azure-openai<0.3
+llama-index-llms-openai<0.3
+sqlalchemy-cratedb
```
Comment (amotl) [inline topic/machine-learning/llama-index/requirements.txt:7]: After cleaning up dependencies, those are apparently all we need to define.

File: topic/machine-learning/llama-index/test.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    return DatabaseAdapter(dburi="crate://crate@localhost:4200")
+
+
+@pytest.fixture(scope="function", autouse=True)
+def init_database(cratedb):
+    """
+    Initialize database.
+    """
+    cratedb.run_sql("DROP TABLE IF EXISTS time_series_data;")
+    cratedb.run_sql((HERE / "init.sql").read_text())
```
[the remark this answers is not in the document]
Comment (amotl) [reply topic/machine-learning/llama-index/test.py]:
```python
from cratedb_toolkit.io.sql import DatabaseAdapter
```
CTK's `DatabaseAdapter` is effectively a little wrapper around a few fundamentals, also using SQLAlchemy, in this spirit very similar to the SQL database wrappers of LangChain and LlamaIndex.

- [LangChain » SQLDatabase: SQLAlchemy wrapper around a database](https://github.com/langchain-ai/langchain/blob/langchain%3D%3D0.3.7/libs/community/langchain_community/utilities/sql_database.py)
- [LlamaIndex » SQLDatabase: A wrapper around the SQLAlchemy engine](https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/utilities/sql_wrapper.py)

[the remark this answers is not in the document]
Comment (amotl) [reply topic/machine-learning/llama-index/test.py]: Note that LangChains also bears another layer on top, the `SQLDatabaseLoader`, that is also pretty generic.

- [SQLDatabaseLoader: Load documents by querying database tables supported by SQLAlchemy](https://github.com/langchain-ai/langchain/blob/langchain%3D%3D0.3.7/libs/community/langchain_community/document_loaders/sql_database.py)

It has been coming from those contributions, broken out of https://github.com/crate-workbench/langchain/pull/1:

- https://github.com/langchain-ai/langchain/pull/16246
- https://github.com/langchain-ai/langchain/pull/16655

**Comment by amotl** [comment]
Do you agree with the changes here, @simonprickett, so we can toggle the badge on the [Build Status page](https://cratedb.com/docs/crate/clients-tools/en/latest/status.html) into green?

File: topic/machine-learning/llama-index/init.sql (hunk 1 above)
Comment (simonprickett) [reply topic/machine-learning/llama-index/init.sql:23]: Let's keep this here as it lines up better with the format of the tutorial on Discourse.

**Review by simonprickett — APPROVED** [review]
I'm OK with this - the tutorial on Discourse will need updating once this has been merged.  I'll take that on.

**Pull request merged**